include/ginkgo/core/base/lin_op.hpp

```diff
@@ -767,6 +767,31 @@ class EnableAbsoluteComputation : public AbsoluteComputable {
 };
 
 
+/**
+ * Mix-in that adds the operation M <- a I + b M for matrix M, identity
+ * operator I and scalars a and b, where M is the calling object.
+ */
+class EnableScaledIdentityAddition {
+public:
+    /**
+     * Scales this and adds another scalar times the identity to it.
+     *
+     * @param a  Scalar to multiply the identity operator by before adding.
+     * @param b  Scalar to multiply this before adding the scaled identity to
+     *   it.
+     */
+    void add_scaled_identity(const LinOp* const a, const LinOp* const b)
+    {
+        GKO_ASSERT_IS_SCALAR(a);
+        GKO_ASSERT_IS_SCALAR(b);
+        add_scaled_identity_impl(a, b);
+    }
+
+private:
+    virtual void add_scaled_identity_impl(const LinOp* a, const LinOp* b) = 0;
+};
```

Comment:
Also in light of #796 I am a bit wary about adding functionality to both Dense and Csr behind a common interface. Can we instead implement the functions separately? Independent of that, you should probably deal with Executor mismatch if you provide this implementation (which necessitates a Mixin and makes the whole interface more difficult to use anyways)

Reply:
I'm treating Dense as a matrix here, yes. But this functionality is needed by Sundials in Dense (it's using gko Dense as a matrix type in addition to using it as a vector type). What would the advantage of having a separate function for Dense be? I'd rather enforce the same interface for the same operation on different formats. Ultimately we have to work with how the code currently is - Dense is both matrix and vector.

Reply:
As for handling executors, that's a good point. I'll add that in the `add_scaled_identity` function.

Reply:
For this to be possible, you need to query the executor of the object, i.e. you need to turn it into a mixin, so for the Identity interface to work, you need two classes "ScaledIdentityAddable" and "EnableScaledIdentityAddition", the former without template paramters for the feature check, the latter as a mixin.

Reply:
Yeah, I tried a couple of things and that seems to be the best solution.